rustls/src/versions.rs

```diff
@@ -37,41 +39,60 @@ pub static ALL_VERSIONS: &[&SupportedProtocolVersion] = &[
 /// This will be [`ALL_VERSIONS`] for now, but gives space in the future
 /// to remove a version from here and require users to opt-in to older
 /// versions.
+//
+// This needs to be kept in sync with the `impl Default for EnabledVersions`.
 pub static DEFAULT_VERSIONS: &[&SupportedProtocolVersion] = ALL_VERSIONS;
 
-#[derive(Debug, Clone)]
-pub(crate) struct EnabledVersions {
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum EnabledVersions {
+    #[cfg(feature = "tls12")]
+    All,
+    Tls13,
     #[cfg(feature = "tls12")]
-    tls12: Option<&'static SupportedProtocolVersion>,
-    tls13: Option<&'static SupportedProtocolVersion>,
+    Tls12,
 }
 
 impl EnabledVersions {
-    pub(crate) fn new(versions: &[&'static SupportedProtocolVersion]) -> Self {
-        let mut ev = Self {
-            #[cfg(feature = "tls12")]
-            tls12: None,
-            tls13: None,
-        };
+    pub(crate) fn new(versions: &[&'static SupportedProtocolVersion]) -> Option<Self> {
```

Comment:
Let's remember the goal of the APIs that take `&[&'static Whatever]` is to make it so that link-time dead-code optimization can eliminate all the dead code. In this case, the intent is that `SupportedProtocolVersion` should have a (crate-private) API that allows us to move the bulk of TLS-1.2-only and TLS-1.3-only logic into methods of `SupportedProtocolVersion`, and eliminate the `if` and `match` on version numbers within the shared code. This way, there would be ~0 TLS-1.2-specific code in a project that doesn't enable TLS 1.2.

But the actual API of `EnabledVersions` and `SupportedProtocolVersion` is oriented towards doing the opposite of that.

We should refactor all of this to orient it with the dead code elimination goal. I bet this would also have the benefit of eliminating a lot of the `#[cfg]` logic based on whether the TLS 1.2 feature flag is enabled.